src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+        // We will dispose it when we are disposed.
+        _channel = CreateChannel();
+
+        _client = new DashboardService.DashboardServiceClient(_channel);
+
+        static Uri GetAddressUri(string variableName, string defaultValue)
+        {
+            try
+            {
+                var uri = Environment.GetEnvironmentVariable(variableName) ?? defaultValue;
+
+                return new Uri(uri);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error parsing URIs from environment variable '{variableName}'.", ex);
+            }
+        }
+
+        GrpcChannel CreateChannel()
+        {
+            var httpHandler = new SocketsHttpHandler
+            {
+                EnableMultipleHttp2Connections = true,
+                KeepAlivePingDelay = TimeSpan.FromSeconds(20),
+                KeepAlivePingTimeout = TimeSpan.FromSeconds(10),
+                KeepAlivePingPolicy = HttpKeepAlivePingPolicy.WithActiveRequests,
+                PooledConnectionIdleTimeout = TimeSpan.FromHours(2)
+            };
+
+            // https://learn.microsoft.com/aspnet/core/grpc/retries
+
+            var methodConfig = new MethodConfig
+            {
+                Names = { MethodName.Default },
+                RetryPolicy = new RetryPolicy
+                {
+                    MaxAttempts = 5,
+                    InitialBackoff = TimeSpan.FromSeconds(1),
+                    MaxBackoff = TimeSpan.FromSeconds(5),
+                    BackoffMultiplier = 1.5,
+                    RetryableStatusCodes = { StatusCode.Unavailable }
+                }
+            };
+
+            // https://learn.microsoft.com/aspnet/core/grpc/diagnostics#grpc-client-logging
+
+            return GrpcChannel.ForAddress(
+                address,
+                channelOptions: new()
+                {
+                    HttpHandler = httpHandler,
+                    ServiceConfig = new() { MethodConfigs = { methodConfig } },
+                    LoggerFactory = _loggerFactory,
+                    ThrowOperationCanceledOnCancellation = true
+                });
+        }
+    }
+
+    private void EnsureInitialized()
+    {
+        var priorState = Interlocked.CompareExchange(ref _state, StateInitialized, comparand: StateNone);
+
+        if (priorState is not StateNone)
+        {
+            ObjectDisposedException.ThrowIf(priorState is StateDisposed, this);
+            return;
+        }
+
+        _connection = Task.Run(() => ConnectAndStayConnectedAsync(_cts.Token), _cts.Token);
+
+        return;
+
+        async Task ConnectAndStayConnectedAsync(CancellationToken cancellationToken)
+        {
+            // Track the number of errors we've seen since the last successfully received message.
+            // As this number climbs, we hold off on 
+            var errorCount = 0;
+
+            await ConnectAsync().ConfigureAwait(false);
```

Comment:
If the initial gRPC call in `ConnectAsync` fails then the connection task will complete with an error and there will never be resource updates. I think the `GetApplicationInformationAsync` call in `ConnectAsync` needs to move somewhere else. The task should just be for watching resources.

Reply:
This method applies to the entire dashboard client, not just gathering resources. Any connection error is grounds to fail. I've refactored the code here to highlight this with clearer method names.

Reply:
What is the user experience of failure?

For example:

* Customer navigates to the Aspire dashboard hosted by a cloud provider
* Transient error in their `GetApplicationInformationAsync` implementation causes connect task to break

What do they see? From what I understand of the current code, data will never load which isn't a good experience. If they need to refresh their browser to unbreak the dashboard then there needs to be a error UI that tells them that.

Reply:
So if `GetApplicationInformationAsync` somehow did throw, the UI would appear with no title and no resources, and with no indication of any problem. I've added a try/catch for this, which at least allows resources to be loaded, and the title remains blank.

Reply:
Ok. It's not ideal but good enough for now. I think a follow up issue is needed for what happens in the UI if the dashboard service is down.

Reply:
https://github.com/dotnet/aspire/issues/1554